README.md
# Promised Retry

A generic promised based retry mechanism. Useful for eg. ensuring an available database or message queue connection

![Linting](https://github.com/voxpelli/node-promised-retry/workflows/Linting/badge.svg)
![Node CI](https://github.com/voxpelli/node-promised-retry/workflows/Node%20CI/badge.svg)
[![Coverage Status](https://coveralls.io/repos/voxpelli/node-promised-retry/badge.svg)](https://coveralls.io/r/voxpelli/node-promised-retry)
[![Known Vulnerabilities](https://snyk.io/test/github/voxpelli/node-promised-retry/badge.svg?targetFile=package.json)](https://snyk.io/test/github/voxpelli/node-promised-retry?targetFile=package.json)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat)](https://github.com/Flet/semistandard)

## Installation

```bash
yarn add promised-retry
```

## Usage

Simple:

```javascript
const Retry = require('promised-retry')
const { Client } = require('pg')

const retryInstance = new Retry({
  try: async () => {
    const db = new Client(require('../config').db);

    db.on('error', () => {
      retryInstance.reset();
      if (channels.length) {
        retryInstance.try();
      }
    });

    await db.connect();
  },
  success: db => {
    db.on('notification', self.processNotification.bind(self));

    channels.forEach(channel => {
      db.query('LISTEN ' + channel);
    });
  },
  end: db => { db.end(); }
});
```

## Syntax

`const retryInstance = new Retry(options);`

## Options

* **try** - the function that will make up the attempt. Should return a Promise that's fulfilled or rejected depending on whether the attempt is a success or failure.
* **success** - a function that's run on an successful attempt. Will be sent the result of the attempt and can return a modified result or a Promise.
* **end** - a function that will be run on the closing of the retry script. Useful for when needing to fix a graceful shutdown of eg. a database. Can return a Promise if it needs to do something that the shutdown needs to wait for.
* **name** – the name of the Retry attempt. Used in eg debugging.
* **setup** – a function that will be run before the first attempt
* **name** – the name of the Retry attempt. Used in eg debugging.
* **retryMin** – minimum delay in milliseconds before a retry. Defaults to `0`.
* **retryBase** – the base of the delay exponent. Defaults to `1.2`.
* **retryExponent** – the maximum exponent of the delay exponent. If retries are higher than `retryExponent`, `retryExponent` will be used rather than the retry number. Defaults to `33` which means on average max delay of 3m 25s.
* **retryDelay** – a function used to calculate the delay. Replaces the default exponent calculation. If it returns `false` the retries will be aborted.
* **retryLimit** – maximum amount of retries. Defaults to unlimited retries.
* **log** – a logger function. Defaults to `console.log()`.

## Methods

* **try** – returns a Promise that will be resolved with the successful attempt or rejected if the retries were aborted due to the result of `options.retryDelay()` or because the retry instance was ended.
* **end** – ends the Retry mechanism completely. Useful for ensuring a graceful shutdown. Returns a Promise that will be resolved when done.
* **reset** – resets the Retry mechanism. Used in response to an event that eg. indicated that the connection the Retry mechanism managed to established has errored out. Will not result in a retry automatically. That has to be done manually if one wants one right away.

## Lint / Test

`npm test` or to watch, install `grunt-cli` then do `grunt watch`


CHANGELOG.md
# Changelog

## 0.5.0 (2022-04-14)

* **Breaking change:** Now targets at least Node.js 14.x
* **Minor:** Removed comments from generated types
* **Minor:** Updated dependencies

## 0.4.0 (2021-02-14)

* **Breaking change:** Now targets at least Node.js 12.x
* **Major addition:** Now comes with TypeScript type definitions
* **Minor internal:** Lots of things, mostly dev dependency updates, but also some refactorings
## 0.3.0 (2018-09-28)

* **Breaking change:** Now targets at least Node.js 8.x
* **Breaking change:** No longer supports a callback on `try()`
* **Breaking change:** No longer throws `AssertionError` on invalid configuration when created

## 0.2.3 (2015-08-20)


### Bug Fixes

* **main:** stopped retrying doesn't mean ending ([7c78b25d](http://github.com/voxpelli/node-promised-retry/commit/7c78b25d5f84dde2e1dfe8e434430d92fd10152e))


### Features

* **main:**
  * allow retries to be limited ([11c5c607](http://github.com/voxpelli/node-promised-retry/commit/11c5c607b5a1c35fd17334b79e122e586b5d03fa))
  * allow retries to be limited ([7bb985c0](http://github.com/voxpelli/node-promised-retry/commit/7bb985c013caceef40216dd809caa225a7900506))


## 0.2.2 (2015-08-05)


### Bug Fixes

* **main:** default retry delay wasn't working ([4f66c045](http://github.com/voxpelli/node-promised-retry/commit/4f66c04568ad1ad73cdfd3630337990a8619e97d))


## 0.2.1 (2015-08-05)


### Bug Fixes

* **main:**
  * if not started, when ending, do nothing ([8be97b23](http://github.com/voxpelli/node-promised-retry/commit/8be97b2373d356a9c2d82638be4f470f4daa432a))
  * more tests + fixed some bugs found ([3883e5ac](http://github.com/voxpelli/node-promised-retry/commit/3883e5ac6c65ad17e807549e6873d62e8613b20a))


## 0.2.0 (2015-08-04)


### Bug Fixes

* **dependencies:**
  * no longer polyfilling Promises ([32349434](http://github.com/voxpelli/node-promised-retry/commit/32349434a01afda660c386e9e5301e4a135d80c1))
  * updated some outdated ones ([395ed70c](http://github.com/voxpelli/node-promised-retry/commit/395ed70cf317c19767f42702c134231ff1246221))
* **main:**
  * added compatibility with newer engines ([c12cdeae](http://github.com/voxpelli/node-promised-retry/commit/c12cdeae2715f9eb184a67c96fff332188d88f3e))
  * documented needed options methods ([02fb5401](http://github.com/voxpelli/node-promised-retry/commit/02fb5401ec6a4998a1b0899b79bb97b9bb5a472f))


### Breaking Changes

* Now requiring iojs or at least node 0.12
 ([32349434](http://github.com/voxpelli/node-promised-retry/commit/32349434a01afda660c386e9e5301e4a135d80c1))



test/main.spec.js
/* eslint-env mocha */
/* eslint-disable n/no-unpublished-require, no-unused-expressions, promise/always-return, promise/prefer-await-to-then */

'use strict';

const chai = require('chai');
const sinon = require('sinon');
const sinonChai = require('sinon-chai');

const should = chai.should();
chai.use(sinonChai);

const Retry = require('..');

const resolved = Promise.resolve();

describe('Retry', () => {
  /** @type {import('sinon').SinonFakeTimers|undefined} */
  let clock;

  /**
   * @template T
   * @param {() => T} condition
   * @param {number} [count]
   * @returns {Promise<T>}
   */
  const repeatUntilCondition = function (condition, count) {
    return new Promise((resolve, reject) => {
      if (count && count > 100) { return reject(new Error('repeatUntilCondition repeated for too long')); }

      const callback = () => {
        resolve(condition() || repeatUntilCondition(condition, count ? count + 1 : 1));
      };
      process.nextTick(() => {
        if (clock) {
          // eslint-disable-next-line promise/catch-or-return, promise/no-callback-in-promise
          resolved.then(callback);
          clock.tick(1000);
        } else {
          callback();
        }
      });
    });
  };

  /**
   * @template T
   * @param {() => T} condition
   * @returns {() => Promise<T>}
   */
  const waitForCondition = function (condition) {
    return () => repeatUntilCondition(condition);
  };

  afterEach(() => {
    clock = undefined;
    sinon.restore();
  });

  describe('main', () => {
    /** @type {import('sinon').SinonStub} */
    let tryStub;
    /** @type {import('sinon').SinonSpy} */
    let successSpy;
    /** @type {import('sinon').SinonSpy} */
    let endSpy;
    /** @type {import('sinon').SinonStub} */
    let retryStub;
    /** @type {import('..')} */
    let retryInstance;

    beforeEach(() => {
      tryStub = sinon.stub();
      successSpy = sinon.spy();
      endSpy = sinon.spy();
      retryStub = sinon.stub().returns(1);

      retryInstance = new Retry({
        'try': tryStub,
        success: successSpy,
        end: endSpy,
        retryDelay: retryStub,
        log: () => {}
      });
    });

    it('should fail if missing a required option', () => {
      // @ts-ignore
      should.Throw(() => new Retry({}), /Promised Retry needs to be provided a "try", "success" and "end" function/);
      // @ts-ignore
      should.Throw(() => new Retry({ 'try': () => {} }), /Promised Retry needs to be provided a "try", "success" and "end" function/);
      // @ts-ignore
      should.Throw(() => new Retry({ 'try': () => {}, success: () => {} }), /Promised Retry needs to be provided a "try", "success" and "end" function/);
    });

    it('should call success on successful try', async () => {
      tryStub.resolves('abc123');

      const result = await retryInstance.try();

      result.should.equal('abc123');

      tryStub.should.have.been.calledOnce;
      successSpy.should.have.been.calledOnce;
      successSpy.should.have.been.calledWith('abc123');
      endSpy.should.not.have.been.called;
      retryStub.should.not.have.been.called;
    });

    it('should do a proper retry', () => {
      clock = sinon.useFakeTimers(Date.now());

      tryStub.onFirstCall().rejects(new Error('foo'));
      tryStub.onSecondCall().rejects(new Error('bar'));
      tryStub.resolves('abc123');

      let fulfilled = false;

      const result = retryInstance.try().then(() => {
        fulfilled = true;

        tryStub.should.have.been.calledThrice;
        successSpy.should.have.been.calledOnce;
        successSpy.should.have.been.calledWith('abc123');
        endSpy.should.not.have.been.calledOnce;
      }).catch(err => {
        fulfilled = true;
        throw err;
      });

      return Promise.resolve()
        .then(waitForCondition(() => tryStub.callCount === 1 && retryStub.callCount === 1))
        .then(waitForCondition(() => tryStub.callCount === 2 && retryStub.callCount === 2))
        .then(waitForCondition(() => fulfilled === true))
        .then(() => result);
    });

    it('should work with default retry', () => {
      clock = sinon.useFakeTimers(Date.now());

      retryInstance = new Retry({
        'try': tryStub,
        success: successSpy,
        end: endSpy,
        log: () => {}
      });

      tryStub.onFirstCall().rejects(new Error('foo'));
      tryStub.onSecondCall().rejects(new Error('bar'));
      tryStub.resolves('abc123');

      let fulfilled = false;

      const result = retryInstance.try().then(() => {
        fulfilled = true;

        tryStub.should.have.been.calledThrice;
        successSpy.should.have.been.calledOnce;
        successSpy.should.have.been.calledWith('abc123');
        endSpy.should.not.have.been.calledOnce;
      }).catch(err => {
        fulfilled = true;
        throw err;
      });

      return Promise.resolve()
        .then(waitForCondition(() => tryStub.callCount === 1))
        .then(waitForCondition(() => tryStub.callCount === 2))
        .then(waitForCondition(() => fulfilled === true))
        .then(() => result);
    });

    it('should only attempt once, despite multiple calls', () => {
      tryStub.resolves('abc123');

      retryInstance.try();
      retryInstance.try();

      return retryInstance.try().then(result => {
        result.should.equal('abc123');

        tryStub.should.have.been.calledOnce;
        successSpy.should.have.been.calledOnce;
        successSpy.should.have.been.calledWith('abc123');
        endSpy.should.not.have.been.called;
        retryStub.should.not.have.been.called;

        return retryInstance.try();
      }).then(result => {
        result.should.equal('abc123');

        tryStub.should.have.been.calledOnce;
        successSpy.should.have.been.calledOnce;
        successSpy.should.have.been.calledWith('abc123');
        endSpy.should.not.have.been.called;
        retryStub.should.not.have.been.called;

        return retryInstance.try();
      });
    });

    it('should call end with successful result on end after success', () => {
      tryStub.resolves('abc123');

      return retryInstance.try()
        .then(() => retryInstance.end())
        .then(() => {
          tryStub.should.have.been.calledOnce;
          successSpy.should.have.been.calledOnce;
          successSpy.should.have.been.calledWith('abc123');
          endSpy.should.have.been.calledOnce;
          endSpy.should.have.been.calledWith('abc123');
          retryStub.should.not.have.been.called;
        });
    });

    it('should abort retries on end call during retries', () => {
      clock = sinon.useFakeTimers(Date.now());

      tryStub.rejects(new Error('foo'));

      retryInstance.try().then(() => {
        process.nextTick(() => {
          throw new Error('try() should not succeed');
        });
      }).catch(() => {});

      let waitTicks = 30;

      return Promise.resolve()
        .then(waitForCondition(() => tryStub.callCount === 1 && retryStub.callCount === 1 ? retryInstance.end() : false))
        .then(waitForCondition(() => --waitTicks === 0))
        .then(() => {
          tryStub.should.have.been.called;
          retryStub.should.have.been.calledOnce;
          successSpy.should.not.have.been.called;
          endSpy.should.have.been.calledOnce;
          endSpy.should.have.been.calledWith();
        });
    });

    it('should not make any attempts if ended before the first one', () => {
      clock = sinon.useFakeTimers(Date.now());

      tryStub.rejects(new Error('foo'));

      let waitTicks = 3;

      return retryInstance.end()
        .then(waitForCondition(() => --waitTicks === 0))
        .then(() => {
          tryStub.should.not.have.been.called;
          retryStub.should.not.have.been.called;
          successSpy.should.not.have.been.called;
          endSpy.should.have.been.calledOnce;
          endSpy.should.have.been.calledWith();
        });
    });

    it('should ignore any attempts if ended', () => {
      tryStub.resolves('abc123');

      return retryInstance.end()
        .then(() => retryInstance.try())
        .then(
          () => { throw new Error('try() should not resolve after an ending'); },
          () => {
            tryStub.should.not.have.been.called;
            retryStub.should.not.have.been.called;
            successSpy.should.not.have.been.called;
            endSpy.should.have.been.calledOnce;
            endSpy.should.have.been.calledWith();
          }
        );
    });

    it('should abort when retry function says so', () => {
      clock = sinon.useFakeTimers(Date.now());

      tryStub.rejects(new Error('foo'));
      retryStub.onFirstCall().returns(1);
      retryStub.returns(false);

      let fulfilled = false;

      const result = retryInstance.try().then(() => {
        fulfilled = true;
        throw new Error('try() should not succeed');
      }, () => {
        fulfilled = true;
      });

      return Promise.resolve()
        .then(waitForCondition(() => tryStub.callCount === 1 && retryStub.callCount === 1))
        .then(waitForCondition(() => fulfilled === true))
        .then(() => {
          tryStub.should.have.been.calledTwice;
          retryStub.should.have.been.calledTwice;

          return result;
        });
    });

    it('should make a new attempt if attempt is requested after a reset', () => {
      tryStub.onFirstCall().resolves('abc123');
      tryStub.resolves('xyz789');

      return retryInstance.try().then(result => {
        result.should.equal('abc123');

        tryStub.should.have.been.calledOnce;
        successSpy.should.have.been.calledOnce;
        successSpy.should.have.been.calledWith('abc123');
        endSpy.should.not.have.been.called;
        retryStub.should.not.have.been.called;

        retryInstance.reset();

        return retryInstance.try();
      }).then(result => {
        result.should.equal('xyz789');

        tryStub.should.have.been.calledTwice;
        successSpy.should.have.been.calledTwice;
        successSpy.should.have.been.calledWith('xyz789');
        endSpy.should.not.have.been.called;
        retryStub.should.not.have.been.called;
      });
    });

    it('should abort retries after limit is reached', () => {
      clock = sinon.useFakeTimers(Date.now());

      tryStub.rejects(new Error('foo'));

      retryInstance = new Retry({
        'try': tryStub,
        success: successSpy,
        end: endSpy,
        retryDelay: retryStub,
        log: () => {},
        retryLimit: 1
      });

      let fulfilled = false;

      const result = retryInstance.try().catch(err => {
        fulfilled = true;

        err.should.be.an('Error');
        err.message.should.equal('Retry limit reached');

        tryStub.should.have.been.calledTwice;
        successSpy.should.not.have.been.called;
        endSpy.should.not.have.been.called;
        retryStub.should.have.been.calledOnce;
      });

      return Promise.resolve()
        .then(waitForCondition(() => tryStub.callCount === 1 && retryStub.callCount === 1))
        .then(waitForCondition(() => tryStub.callCount === 2))
        .then(waitForCondition(() => fulfilled === true))
        .then(() => result);
    });
  });
});


index.js
// @ts-check
/// <reference types="node" />

'use strict';

const {
  ErrorWithCause,
  findCauseByReference,
} = require('pony-cause');

/** @type {() => void} */
const noop = () => {};

class RetryError extends Error {
  /**
   * @param {string} message
   * @param {{ aborted?: boolean }} [options]
   */
  constructor (message, { aborted } = {}) {
    super(message);

    /** @type {string} */
    this.name = this.constructor.name;
    Error.captureStackTrace(this, this.constructor);

    /** @type {boolean} */
    this.aborted = !!aborted;
  }
}

/** @returns {typeof console.log} */
const getDefaultLog = () =>
  // eslint-disable-next-line n/no-process-env
  process.env['NODE_ENV'] === 'production'
    ? () => {}
    // eslint-disable-next-line no-console
    : console.log.bind(console);

/** @typedef {(failures: number) => number|false} RetryDelayCallback */

/**
 * @param {{ retryBase: number, retryExponent: number, retryMin: number }} options
 * @returns {RetryDelayCallback}
 */
const getDefaultRetryDelay = ({ retryBase, retryExponent, retryMin }) => {
  return (retries) => {
    return retryMin + Math.floor(
      1000 *
      Math.pow(retryBase, Math.min(retryExponent, retries)) *
      Math.random()
    );
  };
};

/**
 * @template T
 * @returns {[Promise<T>, (value: T | PromiseLike<T>) => void, (err: Error) => void]}
 */
const resolveablePromise = () => {
  /** @type {(value: T | PromiseLike<T>) => void} */
  let resolver;
  /** @type {(err: Error) => void} */
  let rejecter;

  /** @type {Promise<T>} */
  const resolveable = new Promise((resolve, reject) => {
    resolver = resolve;
    rejecter = reject;
  });

  return [
    resolveable,
    // @ts-ignore
    resolver,
    // @ts-ignore
    rejecter,
  ];
};

/**
 * @typedef RetryOptions
 * @property {string} [name="unknown"]
 * @property {() => void|Promise<void>} [setup]
 * @property {() => Promise<any>} try
 * @property {function(any): any} success
 * @property {function(any): any} end
 * @property {number} [retryMin=0]
 * @property {number} [retryBase=1.2]
 * @property {number} [retryExponent=33]
 * @property {number} [retryLimit]
 * @property {RetryDelayCallback} [retryDelay]
 * @property {(message: string) => void} [log]
 */

// TODO: Convert into a type templated function instead, to be able to derive the resolved type
class Retry {
  /** @param {RetryOptions} options */
  constructor (options) {
    const {
      log = getDefaultLog(),
      name = 'unknown',
      retryMin = 0,
      retryBase = 1.2,
      retryExponent = 33,
      retryDelay,
      ...optionalOptions
    } = options;

    const resolvedOptions = {
      name,
      retryMin,
      retryBase,
      retryExponent,
      retryDelay: retryDelay || getDefaultRetryDelay({ retryMin, retryBase, retryExponent }),
      ...optionalOptions,
    };

    if (!resolvedOptions.try || !resolvedOptions.success || !resolvedOptions.end) {
      throw new Error('Promised Retry needs to be provided a "try", "success" and "end" function');
    }

    /** @protected */
    this.options = resolvedOptions;
    /**
     * @protected
     * @type {(message: string) => void}
     */
    this.log = log;
    /**
     * @protected
     * @type {number}
     */
    this.failures = 0;
    /**
     * @protected
     * @type {NodeJS.Timeout|undefined}
     */
    this.retrying = undefined;
    /**
     * @protected
     * @type {undefined|((err: Error) => void)}
     */
    this.abort = undefined;
    /**
     * @protected
     * @type {Promise<*>|undefined}
     */
    this.promisedResult = undefined;
  }

  /**
   * @protected
   * @param {unknown} err
   * @returns {Promise<any>}
   */
  async _maybeRetry (err) {
    this.log(
      `Failed retry attempt for ${this.options.name}: ` +
      (err instanceof Error ? err.stack : `Unknown error of type: ${typeof err}`)
    );

    // If we're stopped or aborted, then we should not give it a new attempt
    if (
      this.stopped ||
      findCauseByReference(err, RetryError)?.aborted
    ) {
      throw new ErrorWithCause(
        `Failed retry attempt for ${this.options.name}`,
        { cause: err }
      );
    }

    this.failures += 1;
    this.retrying = undefined;
    this.abort = undefined;

    // If we have reached a retry limit, then we should not give it a new attempt
    if (this.options.retryLimit !== undefined && this.failures > this.options.retryLimit) {
      throw new Error('Retry limit reached');
    }

    // TODO: Why this instead of Promise.resolve().then(() => this._try()) ?
    // Otherwise: Try again!
    return new Promise(resolve => {
      process.nextTick(() => {
        resolve(this._try());
      });
    });
  }

  /**
   * @protected
   * @returns {Promise<any>}
   */
  async _throttledAttempt () {
    const [result, resolve, reject] = resolveablePromise();

    // eslint-disable-next-line promise/prefer-await-to-then
    const next = () => resolve(Promise.resolve().then(async () => {
      this.retrying = undefined;
      this.abort = undefined;
      if (this.stopped) {
        throw new Error(`${this.options.name} has been stopped`);
      }
      return this.options.try();
    }));

    if (this.failures) {
      const delay = this.options.retryDelay(this.failures);
      if (delay !== false) {
        this.log(`Retry ${this.failures}: Waiting ${delay} ms to try ${this.options.name} again`);
        this.retrying = setTimeout(next, delay);
        this.abort = reject;
      } else {
        throw new RetryError(`Retries aborted after ${this.failures} attempts`, { aborted: true });
      }
    } else {
      process.nextTick(next);
    }

    return result;
  }

  /**
   * @protected
   * @returns {Promise<any>}
   */
  async _try () {
    try {
      // We use await here to ensure that we can catch any errors and retry rather than forward errors
      const resolvedValue = await this._throttledAttempt();

      this.failures = 0;

      return resolvedValue;
    } catch (err) {
      return this._maybeRetry(err);
    }
  }

  /**
   * @protected
   * @returns {Promise<any>}
   */
  async _createTryPromise () {
    if (this.options.setup) await this.options.setup();

    const result = await this._try();

    this.log(`Successful retry attempt for ${this.options.name}`);

    return this.options.success(result) || result;
  }

  // TODO: When result has been resolved, stop caching the promise
  // TODO: Improve return type!
  /**
   * @param {boolean} [createNew]
   * @returns {Promise<*>}
   */
  async try (createNew = true) {
    if (!this.promisedResult) {
      if (createNew === false || this.stopped) {
        throw new Error('No available instance');
      }
      this.promisedResult = this._createTryPromise();
    }

    return this.promisedResult;
  }

  async end () {
    this.stopped = true;

    if (this.retrying) {
      clearTimeout(this.retrying);
      this.retrying = undefined;
      this.promisedResult = undefined;
      if (this.abort) {
        this.abort(new Error(`Retries of ${this.options.name} ended`));
      }
    }

    const result = await this.try(false).catch(noop);

    this.promisedResult = undefined;

    return this.options.end(result);
  }

  reset () {
    this.promisedResult = undefined;
    this.failures = 0;
  }
}

module.exports = Retry;
